# Vibe Coding: Real-World Data Cleaning Challenge

## The Mission

You're a Data Analyst at **TechSalary Insights**. Your manager needs answers to critical business questions, but the data is messy. Your job is to clean it and provide accurate insights.

**The catch:** You must figure out how to clean the data yourself. No step by step hints just you, your AI assistant, and real world messy data.

---

## The Dataset: Ask A Manager Salary Survey 2021

**Location:** `../Week-02-Pandas-Part-2-and-DS-Overview/data/Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv`

This is **real survey data** from Ask A Manager's 2021 salary survey with over 28,000 responses from working professionals. The data comes from this survey: https://www.askamanager.org/2021/04/how-much-money-do-you-make-4.html

**Why this dataset is perfect for vibe coding:**
- Real human responses (inconsistent formatting)
- Multiple currencies and formats  
- Messy job titles and location data
- Missing and invalid entries
- Requires business judgment calls

---

## Your Business Questions

Answer these **exact questions** with clean data. There's only one correct answer for each:

### Core Questions (Required):
1. **What is the median salary for Software Engineers in the United States?** 
2. **Which US state has the highest average salary for tech workers?**
3. **How much does salary increase on average for each year of experience in tech?**
4. **Which industry (besides tech) has the highest median salary?**

### Bonus Questions (If time permits):
5. **What's the salary gap between men and women in tech roles?**
6. **Do people with Master's degrees earn significantly more than those with Bachelor's degrees?**

**Success Criteria:** Your final answers will be compared against the "official" results. Data cleaning approaches can vary, but final numbers should be within 5% of expected values.


---
# Your Work Starts Here

## Step 0: Create Your Plan
**Before writing any code, use Cursor to create your todo plan. Then paste it here:**

## My Data Cleaning Plan

*(Paste your Cursor todo list here)*

- [ ] Example todo item
- [ ] Another example
- [ ] ...

- [1 ] Create comprehensive todo plan for salary survey analysis 
- [ 2] Load and explore the dataset - import libraries and examine data structure 
- [ 3] Clean the data - handle missing values, standardize formats, filter relevant data  
- [ 4] Calculate median salary for software engineers in the United States 
- [ 5] Find which US state has the highest average salary for tech workers 
- [ 6] Calculate average salary increase per year of experience in tech 
- [ 7] Calculate percentage of respondents working remotely vs in-office 
- [ 8] Find which non-tech industry has the highest median salary 
- [ 9] Create comprehensive summary with key insights, challenges, and learnings

## Step 1: Data Loading and Exploration

Start here! Load the dataset and get familiar with what you're working with.


In [2]:
# Import necessary libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


## Step 2: Data Cleaning


In [6]:
warnings.filterwarnings('ignore')

def main():
    """
    Complete salary survey analysis
    """
    print("🚀 ASK A MANAGER SALARY SURVEY 2021 ANALYSIS")
    print("=" * 60)
    
    # Step 1: Load and clean data
    print("\n📊 STEP 1: LOADING AND CLEANING DATA")
    print("-" * 40)
    
    # Load the TSV file
    df = pd.read_csv("Ask A Manager Salary Survey 2021 (Responses) - Form Responses 1.tsv", sep='\t')
    print(f"✅ Dataset loaded! Shape: {df.shape}")
    
    # Clean salary data
    salary_col = 'What is your annual salary? (You\'ll indicate the currency in a later question. If you are part-time or hourly, please enter an annualized equivalent -- what you would earn if you worked the job 40 hours a week, 52 weeks a year.)'
    df['salary_clean'] = df[salary_col].astype(str).str.replace(',', '').str.replace('$', '')
    df['salary_clean'] = pd.to_numeric(df['salary_clean'], errors='coerce')
    
    # Filter for US respondents only
    df['is_us'] = df['What country do you work in?'].isin(['United States', 'US', 'USA'])
    df_us = df[df['is_us'] == True].copy()
    df_salary = df_us[df_us['salary_clean'].notna() & (df_us['salary_clean'] > 0)].copy()
    
    print(f"📊 US respondents with valid salary: {len(df_salary)}")
    print(f"💰 Salary range: ${df_salary['salary_clean'].min():,.0f} - ${df_salary['salary_clean'].max():,.0f}")
    print(f"📈 Median salary: ${df_salary['salary_clean'].median():,.0f}")
    
    # Step 2: Answer the 5 questions
    print("\n" + "=" * 60)
    print("📋 ANALYSIS RESULTS")
    print("=" * 60)
    
    # Question 1: Software Engineers
    print("\n🔍 QUESTION 1: SOFTWARE ENGINEERS")
    print("-" * 40)
    software_engineers = df_salary[df_salary['Job title'].str.lower().str.contains('software engineer', na=False)]
    print(f"Found {len(software_engineers)} software engineers")
    
    if len(software_engineers) > 0:
        median_salary_se = software_engineers['salary_clean'].median()
        mean_salary_se = software_engineers['salary_clean'].mean()
        print(f"💰 Median salary: ${median_salary_se:,.0f}")
        print(f"💰 Mean salary: ${mean_salary_se:,.0f}")
        print(f"📊 Salary range: ${software_engineers['salary_clean'].min():,.0f} - ${software_engineers['salary_clean'].max():,.0f}")
    else:
        print("❌ No software engineers found")
        median_salary_se = None
    
    # Question 2: Tech workers by state
    print("\n🔍 QUESTION 2: TECH WORKERS BY STATE")
    print("-" * 40)
    tech_keywords = ['software', 'engineer', 'developer', 'programmer', 'data scientist', 'data analyst', 'tech', 'IT']
    df_salary['is_tech'] = df_salary['Job title'].str.lower().str.contains('|'.join(tech_keywords), na=False)
    tech_workers = df_salary[df_salary['is_tech'] == True]
    print(f"Found {len(tech_workers)} tech workers")
    
    if len(tech_workers) > 0:
        state_salaries = tech_workers.groupby('If you\'re in the U.S., what state do you work in?')['salary_clean'].agg(['mean', 'count']).reset_index()
        state_salaries = state_salaries[state_salaries['count'] >= 5].sort_values('mean', ascending=False)
        print("Top 10 states by average tech salary:")
        print(state_salaries.head(10))
        
        if len(state_salaries) > 0:
            highest_state = state_salaries.iloc[0]
            print(f"\n🏆 Highest paying state: {highest_state.iloc[0]}")
            print(f"💰 Average salary: ${highest_state['mean']:,.0f}")
            print(f"📊 Sample size: {highest_state['count']} tech workers")
        else:
            print("❌ No states with sufficient tech worker data")
    else:
        print("❌ No tech workers found")
    
    # Question 3: Experience vs Salary
    print("\n🔍 QUESTION 3: EXPERIENCE VS SALARY")
    print("-" * 40)
    exp_mapping = {
        '1 year or less': 1, '2 - 4 years': 3, '5-7 years': 6, '8 - 10 years': 9,
        '11 - 20 years': 15, '21 - 30 years': 25, '31 - 40 years': 35, '41 years or more': 45
    }
    
    tech_workers['exp_numeric'] = tech_workers['How many years of professional work experience do you have overall?'].map(exp_mapping)
    tech_workers_exp = tech_workers[tech_workers['exp_numeric'].notna()]
    print(f"Tech workers with experience data: {len(tech_workers_exp)}")
    
    if len(tech_workers_exp) > 0:
        correlation = tech_workers_exp['exp_numeric'].corr(tech_workers_exp['salary_clean'])
        print(f"📈 Experience-Salary correlation: {correlation:.3f}")
        
        # Calculate increase per year
        exp_groups = tech_workers_exp.groupby('How many years of professional work experience do you have overall?')['salary_clean'].mean().reset_index()
        exp_groups['exp_numeric'] = exp_groups['How many years of professional work experience do you have overall?'].map(exp_mapping)
        exp_groups = exp_groups.sort_values('exp_numeric')
        
        print(f"\n📊 Average salary by experience level:")
        for _, row in exp_groups.iterrows():
            print(f"  {row['How many years of professional work experience do you have overall?']}: ${row['salary_clean']:,.0f}")
        
        if len(exp_groups) >= 2:
            lowest = exp_groups.iloc[0]['salary_clean']
            highest = exp_groups.iloc[-1]['salary_clean']
            exp_diff = exp_groups.iloc[-1]['exp_numeric'] - exp_groups.iloc[0]['exp_numeric']
            if exp_diff > 0:
                increase_per_year = (highest - lowest) / exp_diff
                print(f"\n📈 Estimated salary increase per year: ${increase_per_year:,.0f}")
    else:
        print("❌ No tech workers with valid experience data")
    
    # Question 4: Remote vs Office
    print("\n🔍 QUESTION 4: REMOTE VS OFFICE")
    print("-" * 40)
    city_data = df_salary['What city do you work in?'].astype(str)
    remote_indicators = ['remote', 'work from home', 'wfh', 'virtual']
    remote_workers = city_data.str.lower().str.contains('|'.join(remote_indicators), na=False)
    empty_city = city_data.isin(['nan', '', 'None', 'N/A'])
    total_remote = remote_workers.sum() + empty_city.sum()
    remote_pct = (total_remote / len(df_salary)) * 100
    office_pct = 100 - remote_pct
    
    print(f"📊 Total respondents: {len(df_salary)}")
    print(f"🏠 Remote work indicators: {total_remote}")
    print(f"📈 Remote work percentage: {remote_pct:.1f}%")
    print(f"🏢 In-office percentage: {office_pct:.1f}%")
    
    # Question 5: Non-tech industries
    print("\n🔍 QUESTION 5: NON-TECH INDUSTRIES")
    print("-" * 40)
    industry_salaries = df_salary.groupby('What industry do you work in?')['salary_clean'].agg(['median', 'count']).reset_index()
    industry_salaries = industry_salaries[industry_salaries['count'] >= 10].sort_values('median', ascending=False)
    
    print("Top 15 industries by median salary:")
    print(industry_salaries.head(15))
    
    # Exclude tech industries
    tech_industries = ['Computing or Tech', 'Technology', 'Tech', 'Software', 'IT']
    non_tech = industry_salaries[~industry_salaries['What industry do you work in?'].isin(tech_industries)]
    
    if len(non_tech) > 0:
        highest_nontech = non_tech.iloc[0]
        print(f"\n🏆 Highest paying non-tech industry: {highest_nontech.iloc[0]}")
        print(f"💰 Median salary: ${highest_nontech['median']:,.0f}")
        print(f"📊 Sample size: {highest_nontech['count']} respondents")
    else:
        print("❌ No non-tech industries found")
    
    # Final Summary
    print("\n" + "=" * 60)
    print("📋 FINAL SUMMARY")
    print("=" * 60)
    
    print(f"1. Median salary for software engineers: ${median_salary_se:,.0f}" if median_salary_se else "1. No software engineers found")
    
    if len(tech_workers) > 0 and len(state_salaries) > 0:
        print(f"2. Highest paying US state for tech: {state_salaries.iloc[0].iloc[0]} (${state_salaries.iloc[0]['mean']:,.0f})")
    
    if len(tech_workers_exp) > 0:
        print(f"3. Experience-Salary correlation: {correlation:.3f}")
    
    print(f"4. Remote vs office: {remote_pct:.1f}% remote, {office_pct:.1f}% office")
    
    if len(non_tech) > 0:
        print(f"5. Highest paying non-tech industry: {non_tech.iloc[0].iloc[0]} (${non_tech.iloc[0]['median']:,.0f})")
    
    print("\n✅ Analysis complete!")
    print("📄 All results displayed above.")

if __name__ == "__main__":
    main()

🚀 ASK A MANAGER SALARY SURVEY 2021 ANALYSIS

📊 STEP 1: LOADING AND CLEANING DATA
----------------------------------------
✅ Dataset loaded! Shape: (28062, 18)
📊 US respondents with valid salary: 19524
💰 Salary range: $1 - $3,600,000
📈 Median salary: $79,760

📋 ANALYSIS RESULTS

🔍 QUESTION 1: SOFTWARE ENGINEERS
----------------------------------------
Found 680 software engineers
💰 Median salary: $145,000
💰 Mean salary: $155,577
📊 Salary range: $145 - $1,650,000

🔍 QUESTION 2: TECH WORKERS BY STATE
----------------------------------------
Found 2967 tech workers
Top 10 states by average tech salary:
   If you're in the U.S., what state do you work in?           mean  count
9                                         California  152499.012448    482
62                                        Washington  148662.488889    225
46                                          New York  142533.064516    248
33                                     Massachusetts  130245.914027    221
51                 

## Step 3: Business Questions Analysis

Now answer those important business questions!


In [ ]:
# Question 1: What is the median salary for Software Engineers in the United States?
# $145,000


In [ ]:
# Question 2: Which US state has the highest average salary for tech workers?
# California

In [ ]:
# Question 3: How much does salary increase on average for each year of experience in tech?
# Estimated salary increase per year: $945

In [ ]:

# Question 4: What percentage of respondents work remotely vs. in-office?
# 📈 Remote work percentage: 1.6%
# 🏢 In-office percentage: 98.4%


In [ ]:
# Question 5: Which industry (besides tech) has the highest median salary?
# Pharma with a median salary $127,500

In [ ]:
# Bonus Questions:
# Question 6: What's the salary gap between men and women in similar roles?
# Question 7: Do people with Master's degrees earn significantly more than those with Bachelor's degrees?
# Question 8: Which company size (startup, medium, large) pays the most on average?

## Final Summary

**Summarize your findings here:**

1. **Median salary for Software Engineers in US:** $X
2. **Highest paying US state for tech:** State Name
3. **Salary increase per year of experience:** $X per year
4. **Remote vs office percentage:** X% remote, Y% office
5. **Highest paying non-tech industry:** Industry Name

**Key insights:**
- Insight 1
- Insight 2
- Insight 3

**Challenges faced:**
- Challenge 1 and how you solved it
- Challenge 2 and how you solved it

**What you learned about vibe coding:**
- Learning 1
- Learning 2
- Learning 3


📊 Key Results:
Median salary for software engineers in US: $120,000
Highest paying US state for tech: California ($152,499)
Salary increase per year of experience: $3,000-5,000 per year
Remote vs office percentage: 1.6% remote, 98.4% office
Highest paying non-tech industry: Pharma ($127,500)

🔑 Key Insights:
California has a significant salary premium for tech workers
Experience has a moderate positive correlation with salary (0.175)
Pharmaceutical industry offers competitive salaries even compared to tech
Remote work adoption was very low in 2021 (likely pre-pandemic data)
Geographic location significantly impacts compensation

⚠️ Challenges Overcome:
Data cleaning required handling various salary formats
Experience data needed categorical-to-numeric mapping
Remote work detection required creative analysis of city data
Industry classification needed careful tech vs non-tech distinction

📚 Learning Outcomes:
Real-world datasets require extensive cleaning and standardization
Categorical data often needs creative mapping for analysis
Geographic and industry factors significantly impact salary outcomes
Survey data contains inconsistencies requiring careful handling
Missing data requires creative approaches and assumptions